docs/references/astro/hybrid-rendering.mdx

```diff
@@ -0,0 +1,59 @@
+---
+title: Astro hybrid rendering
+description: Clerk's Astro SDK supports Astro's on-demand rendering output modes (`server` and `hybrid`).
+---
+
+Astro's on-demand rendering output modes (`server` and `hybrid`) allow you to pre-render certain pages while keeping others server-rendered. The Clerk Astro SDK supports these output modes out-of-the-box - no additional configuration is required. However, you may need to make some adjustments to your code to ensure that Clerk's control components are rendered correctly in hybrid mode.
+
+> [!TIP]
+> What are control components?
+>
+> Control components manage authentication-related behaviors in your application. They handle tasks such as controlling content visibility based on user authentication status, managing loading states during authentication processes, and redirecting users to appropriate pages. A common example is the [`<SignedIn>`](/docs/components/control/signed-in) component, which allows you to conditionally render content only when a user is authenticated.
+
+## Server output mode
+
+In `server` output mode, pages and control components are server-rendered by default, but you can opt-in to pre-rendering specific pages by adding `export const prerender = true` to the page. When you opt-in to pre-rendering a page, you must add `isStatic={true}` to any control components used on that page. This specifies that the component should use the client-side version, which relies on [client nanostores](/docs/references/astro/auth-store).
+
+The following example shows how to opt-in to pre-rendering a page and specify that the control components used on that page should use the client-side version.
+
+```astro {{ filename: 'src/pages/index.astro' }}
+---
+export const prerender = true
+---
+
+<SignedIn isStatic={true}> You are signed in! </SignedIn>
+```
+
+## Hybrid output mode
+
+In `hybrid` output mode, pages and control components are pre-rendered by default, but you can opt-out of pre-rendering for specific pages by adding `export const prerender = false`. When you opt-out of pre-rendering a page, you must add the `isStatic={false}` prop to any control components used on that page. This specifies that the component should use the server-side version which relies on the [locals](/docs/references/astro/locals) injected by the middleware.
+
+The following example shows how to opt-out of pre-rendering a page and specify that the control components used on that page should use the server-side version.
+
+```astro {{ filename: 'src/pages/index.astro' }}
+---
+export const prerender = false
+---
+
+<SignedIn isStatic={false}> You are signed in! </SignedIn>
+```
+
+## Styling considerations
+
+When using the client-side version of control components (i.e., when `isStatic={true}`), be aware of the wrapper element. For example, the `<SignedIn>` component is wrapped in a custom element. This wrapper may affect your styling, so you might need to adjust your CSS selectors accordingly.
```

Comment:
I am confused by this explanation - could you maybe show me what you mean? Show me how it could affect styling - like what the issue could be here.

Reply:
Yeah it indeed is confusing now that I read it again 😆 So let me try to explain it:

If you opt-out of prerendering in `hybrid` mode:

```astro
---
import { SignedIn } from '@clerk/astro/components'

export const prerender = false // opt-out of prerendering
---

<SignedIn>
  <div>Hello</div>
</SignedIn>
```

That will be rendered as 

```html
<div>Hello</div>
```

and if you don't opt-out, which is the default behavior, that component above will be rendered with an extra element wrapping them (because we manually hide the contents using `hidden` attribute):

```html
<clerk-signed-in>
  <div>Hello</div>
</clerk-signed-in>
```

That `<clerk-signed-in>` is a [custom element](https://developer.mozilla.org/en-US/docs/Web/API/Web_components/Using_custom_elements#using_the_lifecycle_callbacks%23sect4), not an Astro component.

For the styling issue, take a look at this example:

```astro
<div class="flex items-center">
  <SignedIn isStatic={true}>
    <span>User</span>
  </SignedIn>
  <span>Always visible</span>
</div>
```

The rendered output of that is

```astro
<div class="flex items-center">
  <clerk-signed-in>
    <span>User</span>
  </clerk-signed-in>
  <span>Always visible</span>
</div>
```

The `clerk-signed-in` wrapper isn't a flex item by default, potentially misaligning the content. To fix this, they would need to apply flex properties to the `<SignedIn>` component:

```astro
<div class="flex items-center">
  <SignedIn isStatic={true} class="flex items-center">
    <span>User</span>
  </SignedIn>
  <span>Always visible</span>
</div>
```